Add an authentication overview doc (#15735)
aspnet/AspNetCore.Docs: aspnetcore/security/authentication/index.md

* Authenticating a user.
* Responding when an unauthenticated user tries to access a restricted resource.

The registered authentication handlers and their configuration options, are called "schemes".

Authentication schemes are specified by registering authentication services in `Startup.ConfigureServices`. This is done by:

* Calling a scheme-specific extension method after a call to `services.AddAuthentication` (such as `AddJwtBearer` or `AddCookie`, for example). These extension methods use [AuthenticationBuilder.AddScheme](xref:Microsoft.AspNetCore.Authentication.AuthenticationBuilder.AddScheme*) to register schemes with appropriate settings.
* Less commonly, by calling [AuthenticationBuilder.AddScheme](xref:Microsoft.AspNetCore.Authentication.AuthenticationBuilder.AddScheme*) directly.

For example, the following code registers authentication services and handlers for cookie and JWT bearer authentication schemes:

```csharp
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)

>     .AddJwtBearer(options => Configuration.Bind("JwtSettings", options))

Reviewer: This example would be clearer if the scheme name was also passed into AddJwtBearer and AddCookie.
```suggestion
    .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options => Configuration.Bind("JwtSettings", options))
```

Author: Sure, I'll update the sample to look like this:

```csharp
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(JwtBearerDefaults.AuthenticationScheme, options => Configuration.Bind("JwtSettings", options))
    .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, options => Configuration.Bind("CookieSettings", options));
```

    .AddCookie(options => Configuration.Bind("CookieSettings", options));
```

The `AddAuthentication` parameter `JwtBearerDefaults.AuthenticationScheme` is the name of the scheme to use by default if a specific one isn't requested when authenticating or authorizing in the application. In the preceding code, JWT bearer authentication is used by default.

> In some cases, the call to `AddAuthentication` is automatically made by other extension methods. For example, when using [ASP.NET Core Identity](xref:security/authentication/identity),`AddAuthentication` is called. If multiple schemes are used, authorization policies (or authorization attributes) can [specify the authentication scheme (or schemes)](xref:security/authorization/limitingidentitybyscheme) they depend on to authenticate the user. In the example above, the cookie authentication scheme could be used by specifying its name (`CookieAuthenticationDefaults.AuthenticationScheme` by default, though a different name could be provided when calling `AddCookie`).

Reviewer: ```suggestion
If multiple schemes are used, authorization policies (or authorization attributes) can [specify the authentication scheme (or schemes)](xref:security/authorization/limitingidentitybyscheme) they depend on to authenticate the user. In the example above, the cookie authentication scheme could be used by specifying its name (`CookieAuthenticationDefaults.AuthenticationScheme` by default, though a different name could be provided when calling `AddCookie`).

In some cases, the call to `AddAuthentication` is automatically made by other extension methods. For example, when using [ASP.NET Core Identity](xref:security/authentication/identity),`AddAuthentication` is called internally. 
```

* An authentication handler. 
* Options for configuring that specific instance of the handler. 
 
Schemes are useful as a mechanism for easily referring to the authentication, challenge, and forbid behaviors of the associated handler. For example, an authorization policy can specify by name which authorization scheme (or schemes) should be used to authenticate the user. When configuring authentication, it's common to specify the default authentication scheme. The default scheme is used unless a resource requests a specific scheme. It's also possible to:

* Specify different default schemes to use for authenticate, challenge, and forbid actions.
* Combine multiple schemes into one using [policy schemes](xref:security/authentication/policyschemes).

### Authentication handler

An authentication handler:

* Is a type that implements the behavior of a scheme.
* Is derived from <xref:Microsoft.AspNetCore.Authentication.IAuthenticationHandler> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationHandler>.
* Has the primary responsibility to authenticate users.

Based on the authentication scheme's configuration and the incoming request context, authentication handlers:

> * Construct <xref:Microsoft.AspNetCore.Authentication.AuthenticationTicket> objects representing the user's identity if authentication is successful.

Reviewer: This is the only place I can think to put an HttpContext.AuthenticateAsync reference. Do we need a dedicated Authenticate section to mirror Challenge and Forbid?

Author: Yea, probably an 'Authenticate' section is a good idea (to mirror Challenge and Forbid). I'll add one.

Author: Ok, I've added an Authenticate section. @Tratcher, can you take a quick look?

* Return 'no result' or 'failure' if authentication is unsuccessful.
* Have methods for challenge and forbid actions for when users attempt to access resources:
  * They are unauthorized to access (forbid).
  * When they are unauthenticated (challenge).

### Challenge

> An authentication challenge is the action an authentication scheme takes when an unauthenticated user requests an endpoint that requires authentication. An authentication challenge is issued, for example, on a request to a restricted resource. Authentication challenge examples include:

Reviewer: ```suggestion
An authentication challenge is invoked by Authorization when an unauthenticated user requests an endpoint that requires authentication. An authentication challenge is issued, for example, when an anonymous user requests a restricted resource or clicks on a login link. Authorization invokes a challenge using the specified authentication scheme(s), or the default if none is specified. See HttpContext.ChallengeAsync. Authentication challenge examples include:
```

* A cookie authentication scheme redirecting the user to a login page.
* A JWT bearer scheme returning a 401 result with a `www-authenticate: bearer` header.

A challenge action should let the user know what authentication mechanism to use to access the requested resource.

### Forbid

> An authentication handler's forbid action is used when an authenticated user attempts to access a resource they are not permitted to access. Authentication forbid examples include:

Reviewer: ```suggestion
An authentication scheme's forbid action is called by Authorization when an authenticated user attempts to access a resource they are not permitted to access. See HttpContext.ForbidAsync. Authentication forbid examples include:
```

---
title: Overview of ASP.NET Core Authentication
author: mjrousos
description: Learn about authentication in ASP.NET Core.
ms.author: riande
ms.custom: mvc
ms.date: 11/22/2019
uid: security/authentication/index
---
# Overview of ASP.NET Core authentication

By [Mike Rousos](https://github.com/mjrousos)

Authentication is the process of determining a user's identity. [Authorization](xref:security/authorization/introduction) is the process of determining whether a user has access to a resource. In ASP.NET Core, authentication is handled by authentication [middleware](xref:fundamentals/middleware/index). The authentication middleware uses registered authentication handlers to complete authentication-related actions. Examples of authentication-related actions include:

> 

Reviewer 2: I'm not sure how important the semantics are here.  But technically, authentication is handled by the `IAuthenticationService` which is what uses registered handlers.  The middleware is responsible for 2 things: if a default authenticate scheme is set, it is responsible for setting HttpContext.User.  Also its responsible for routing challenge/requests which are specific for remote authentication flows.  Its probably worth being accurate and spelling out this distinction here.

Reviewer 2: So while most of our implementations of handlers do require our authentication middleware to be present for a mix of not always overlapping reasons, its not actually a requirement at the lowest authentication abstraction layers (IAuthenticationHandler/Schemes).

Reviewer: > So while most of our implementations of handlers do require our authentication middleware to be present for a mix of not always overlapping reasons, its not actually a requirement at the lowest authentication abstraction layers (IAuthenticationHandler/Schemes).

That distinction is not meaningful to most users. Telling them to always add the middleware is simpler.

Author: I can be more specific about how the handlers are used: 
"In ASP.NET Core, authentication is handled by the `IAuthenticationService` (which is called by authentication middleware). The authentication service uses registered handlers..."

I agree with @Tratcher, though, that for this doc is probably best to just leave it at middleware calls service which calls handlers. As long as developers know those different parts exist, they can learn more about them (and how to use them separate from one another) later.

* Authenticating a user.
* Responding when an unauthenticated user tries to access a restricted resource.

The registered authentication handlers and their configuration options, are called "schemes".

Authentication schemes are specified by registering authentication services in `Startup.ConfigureServices`. This is done by:

> * Calling [AuthenticationBuilder.AddScheme](xref:Microsoft.AspNetCore.Authentication.AuthenticationBuilder.AddScheme*)

Reviewer 3: Should we put the more common way first?

Author: My thought was to put the lower level way first (so there's less behind the scenes magic) and then call out the more common way.

Author: I've modified this to put the more common way first.

> * Calling a scheme-specific extension methods that call `AddScheme` automatically with appropriate settings. This approach is more common than calling `AuthenticationBuilder.AddScheme`.

Author: I've modified this to put the more common way first.

For example, the following code registers authentication services and handlers for cookie and JWT bearer authentication schemes:

```csharp
services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(options => Configuration.Bind("JwtSettings", options))
    .AddCookie(options => Configuration.Bind("CookieSettings", options));
```

> The `AddAuthentication` parameter `JwtBearerDefaults.AuthenticationScheme` is the name of the scheme to use by default if a specific one isn't requested. In the preceding code, JWT bearer authentication is used by default.

Reviewer 3: Might want to expand on this. Perhaps add cookie with a specific scheme name to show the contrast?

Author: Good idea. I'll do that.

In some cases, the call to `AddAuthentication` is automatically made by other extension methods. For example, when using [ASP.NET Core Identity](xref:security/authentication/identity)),`AddAuthentication` is called. If multiple schemes are used, authorization policies (or authorization attributes) can [specify the authentication scheme (or schemes)](xref:security/authorization/limitingidentitybyscheme) they depend on to authenticate the user.

> Authentication middleware is added in `Startup.Configure` by calling the <xref:Microsoft.AspNetCore.Builder.AuthAppBuilderExtensions.UseAuthentication*> extension method on the app's `IApplicationBuilder`. This registers the  middleware which uses the previously registered authentication schemes. Call `UseAuthentication` before any middleware that depends on users being authenticated. When using endpoint routing, the call to `UseAuthentication` should go:

Reviewer 3: "must go" not "should go"

Author: Agreed. I'll change it.

Author: Fixed

> Authentication middleware is added in `Startup.Configure` by calling the <xref:Microsoft.AspNetCore.Builder.AuthAppBuilderExtensions.UseAuthentication*> extension method on the app's `IApplicationBuilder`. This registers the  middleware which uses the previously registered authentication schemes. Call `UseAuthentication` before any middleware that depends on users being authenticated. When using endpoint routing, the call to `UseAuthentication` should go:

Author: Fixed

* After `UseRouting`, so that route information is available for authentication decisions.
* Before `UseEndpoints`, so that users are authenticated before accessing the endpoints.

## Authentication Concepts

### Authentication scheme

> An authentication scheme is:

Reviewer 3: @HaoK how do you feel about this? To me a scheme is just a dictionary key for an authentication handler. I'm not sure I'd say the scheme is the way of authenticating the user or responding if there's an unauthenticated or forbidden user, that's the handler.

Author: The bullets here are just referring to the different methods the authentication handler supplies (Authenticate, Challenge, Forbid). As you say, a scheme is just a key for a handler (along with configuration, since the same handler might be registered multiple times as different schemes) but I want to define scheme, specifically, since it's used fairly often (and authentication handler is not) and I know some customers have been confused about what the term means.

Maybe the better approach is to define an authentication scheme as the *name* of an authentication mechanism? Something like: "An authentication scheme is a name which corresponds to an authentication handler and options for configuring that specific instance of the handler. Schemes are useful as a mechanism for easily referring to the authentication, challenge, and forbid behaviors of the associated handler."

Thoughts?

Reviewer 2: Yes that's exactly how I would describe it:

Authentication Scheme = the logical name of one particular instance of an authentication handler.  This also helps when discussing virtual schemes and how that opens up redirection specific authentication actions to different handlers/other schemes.

* A way of authenticating a user.
* How the app responds if a user isn't authenticated.

In ASP.NET Core, an authentication scheme consists of:

* An authentication handler.
* Options for configuring that specific instance of the handler.

Authentication schemes have names so that they can be referenced in the app. For example, an authorization policy that specifies which authorization scheme (or schemes) should be used to authenticate the user. When configuring authentication, it's common to specify the default authentication scheme. The default scheme is used unless a resource requests a specific scheme. It's also possible to:

* Specify different default schemes to use for authenticate, challenge, and forbid actions.
* Combine multiple schemes into one using [policy schemes](xref:security/authentication/policyschemes).

### Authentication handler

An authentication handler:

* Is a type that implements the behavior of a scheme.

> * Is derived from <xref:Microsoft.AspNetCore.Authentication.IAuthenticationHandler>.

Reviewer: Or AuthenticationHandler

* Has the primary responsibility to authenticate users.

Based on the authentication scheme's configuration and the incoming request context, authentication handlers:

* Construct <xref:Microsoft.AspNetCore.Authentication.AuthenticationTicket> objects representing the user's identity if authentication is successful.
* Return 'no result' or 'failure' if authentication is unsuccessful.
* Have methods for challenge and forbid actions for when users attempt to access resources:

>   * They are unauthorized to access.

Reviewer: (Forbid)

>   * When they are unauthenticated.

Reviewer: (Challenge)

### Challenge

An authentication challenge is the action an authentication scheme takes when an unauthenticated user requests an endpoint that requires authentication. An authentication challenge is issued, for example, on a request to a restricted resource. Authentication challenge examples include:

* Cookie authentication scheme redirecting the user to a login page.

> * The JWT bearer scheme returning a 401 result.

Reviewer: and a `www-authenticate: bearer` header.

Author: Good point. I'll add this.

> A challenge action should let the user know that they need to be authenticated to access a requested resource.

Reviewer: ```suggestion
A challenge action should let the user know what authentication mechanism to use to access the requested resource.
```

### Forbid

> An authentication handler's forbid action is used when an authenticated user attempts to access a resource they are not permitted to access. A forbid action typically returns a 403 result. In some custom authentication schemes, forbid could result in:

Reviewer: Keep this symmetrical with Challenge:
```suggestion
An authentication handler's forbid action is used when an authenticated user attempts to access a resource they are not permitted to access. Authentication forbid examples include:
* Cookie authentication scheme redirecting the user to a page indicating access was forbidden.
* The JWT bearer scheme returning a 403 result.
```

* A redirect to a page where the user can request access to the resource.

> For example, the user my request access from an admin. A forbid action should let the user know:

Reviewer 3: can, not should. Arguably 401/403 don't let the user know, it's the app response to the status codes that does.

Author: Makes sense. I'll adjust this.
